### PR TITLE
Multiple observe

### LINKED ISSUE #43: Observe multiple peers
we should be able to observe multiple peers, to support semantics like "block is considered committed if it's done on >50% peers"

### PR DESCRIPTION
**Pull request by SamYuan1990**
resolve #43

### CODE REVIEW

**Review by guoger — CHANGES_REQUESTED** [review]
based on current implementation, it should be possible to print committed blocks *out of order*, right? i.e.
```
block 1
block 3
block 2
block 4
...
```

File: pkg/infra/observer.go [hunk 1]
```diff
@@ -53,7 +84,31 @@ func (o *Observer) Start(N int, errorCh chan error, finishCh chan struct{}, now
...
+			blockcollection.Collection[fb.FilteredBlock.Number] = o.countDown - 1
+			if o.countDown == 1 {
+				fmt.Printf("Time %8.2fs\tBlock %6d\tTx %6d\t \n", time.Since(now).Seconds(), fb.FilteredBlock.Number, len(fb.FilteredBlock.FilteredTransactions))
+			}
+		}
+		// release lock
+		blockcollection.lock.Unlock()
 		n = n + len(fb.FilteredBlock.FilteredTransactions)
-		fmt.Printf("Time %8.2fs\tBlock %6d\tTx %6d\n", time.Since(now).Seconds(), fb.FilteredBlock.Number, len(fb.FilteredBlock.FilteredTransactions))
+		o.logger.Debugf("Time %8.2fs\tBlock %6d\tTx %6d\t Address %s\n", time.Since(now).Seconds(), fb.FilteredBlock.Number, len(fb.FilteredBlock.FilteredTransactions), o.Address)
```
[the remark this answers is not in the document]
Comment (SamYuan1990) [reply pkg/infra/observer.go:112]: With multiple observes,  tape logs/outputs as
1. Block receive from specific peer. -> log
1. Block committed at "all" peers. -> output

what do you think?
I suppose it is too mess as 

1. Block receive from specific peer 0. -> output
1. Block receive from specific peer 1. -> output
...
1. Block receive from specific peer n. -> output
1. Block committed at "all" peers. -> output

File: pkg/infra/observer.go [hunk 2]
```diff
@@ -2,19 +2,50 @@ package infra
...
 	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 )
 
+type BlockCollection struct {
+	Collection map[uint64]int
+	lock       sync.Mutex
+}
+
+type Observers struct {
```
[the remark this answers is not in the document]
Comment (SamYuan1990) [reply pkg/infra/observer.go:18]: keep same function and behavior as `CreateProposers` and `CreateBroadcasters`

File: pkg/infra/observer.go [hunk 3]
```diff
@@ -2,19 +2,50 @@ package infra
...
+	"sync"
 	"time"
 
 	"github.com/hyperledger/fabric-protos-go/peer"
 	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 )
 
+type BlockCollection struct {
+	Collection map[uint64]int
```
[the remark this answers is not in the document]
Comment (SamYuan1990) [reply pkg/infra/observer.go:14]: what do you think for https://github.com/fanliao/go-concurrentMap this one as a concurrent map to store the collection?

[the remark this answers is not in the document]
Comment (SamYuan1990) [reply pkg/infra/observer.go:14]: or sync.Map?

[the remark this answers is not in the document]
Comment (guoger) [reply pkg/infra/observer.go:14]: i don't think there will be significant improvement, but maybe you could demonstrate with a benchmark test?

[the remark this answers is not in the document]
Comment (SamYuan1990) [reply pkg/infra/observer.go:14]: ok, if possible, I will add a benchmark test for it. the point is to avoid making another wheel of concurrentMap.

**Comment by SamYuan1990** [comment]
https://dev.azure.com/guojiannan1101/Tape/_build/results?buildId=568&view=logs&j=507b79f5-c24b-5a50-206f-843997dc3f64&t=1b04bba2-771b-5e51-1e42-e0cb4bc865e4&l=1137 

log as output sample for ref.

File: .gitignore [hunk 4]
```diff
@@ -11,6 +11,3 @@
 
 # Output of the go coverage tool, specifically when used with LiteIDE
 *.out
-
-# Files from IDE
-*.idea/
```
Comment (guoger) [inline .gitignore:16]: why is this removed?

File: README.md [hunk 5]
```diff
@@ -43,14 +43,15 @@ endorsers:
...
-  tls_ca_cert: /path/to/peer2/tls/ca/cert
+committers:
+  - addr: localhost:7051
+    tls_ca_cert: /path/to/peer2/tls/ca/cert
 orderer:
   addr: localhost:7050
   tls_ca_cert: /path/to/orderer/tls/ca/cert
 channel: mychannel
 chaincode: basic
+numberofagreement: 1
```
Comment (guoger) [inline README.md:54]: threshold or commitThreshold?

File: pkg/infra/BlockCollector.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package infra
```
Comment (guoger) [inline pkg/infra/BlockCollector.go:1]: file name should `block_collector.go`

File: pkg/infra/BlockCollector.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+			}
+			// update map
+			bc.Registry[block] = i
+		}
+	} else {
+		// write map
+		bc.Registry[block] = bc.Observers - 1
+	}
+	return false
+}
```
Comment (guoger) [inline pkg/infra/BlockCollector.go:42]:
```go
package infra

import "sync"

type BlockCollector struct {
	sync.Mutex

	threshold int
	registry  map[uint64]int
}

func NewBlockCollector(threshold int) *BlockCollector {
	registry := make(map[uint64]int)
	return &BlockCollector{
		threshold: threshold,
		registry:  registry,
	}
}

func (bc *BlockCollector) Commit(block uint64) (committed bool) {
	if bc.threshold == 1 {
		return true
	}

	bc.Lock()
	defer bc.Unlock()

	remaining, ok := bc.registry[block]
	if !ok {
		bc.registry[block] = bc.threshold - 1
		return false
	}

	if remaining == 1 {
		delete(bc.registry, block)
		return true
	}

	bc.registry[block]--
	return false
}
```
- no need to export internal members
- less nested if-else
- more meaningful var names

File: pkg/infra/BlockCollector_test.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+			time.Sleep(time.Millisecond * 100)
+			Expect(len(instance.Registry)).Should(Equal(1))
+			i, ok := instance.Registry[1]
+			Expect(ok).To(BeTrue())
+			Expect(i).To(Equal(1))
+			instance.Commit(uint64(1))
+			Expect(len(instance.Registry)).Should(Equal(0))
+		})
+	})
+})
```
Comment (guoger) [inline pkg/infra/BlockCollector_test.go:56]:
```go
package infra_test

import (
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
	"sync"
	"tape/pkg/infra"
)

var _ = Describe("BlockCollector", func() {

	Context("Commit", func() {
		It("should return true when observer is 1", func() {
			instance := infra.NewBlockCollector(1)
			Expect(instance.Commit(uint64(1))).To(BeTrue())
		})

		It("Should added when the block committed at 1st time", func() {
			instance := infra.NewBlockCollector(3)
			Expect(instance.Commit(uint64(1))).To(BeFalse())
			Expect(instance.Commit(uint64(1))).To(BeFalse())
			Expect(instance.Commit(uint64(1))).To(BeTrue())
		})

		It("Should supports multiple running", func() {
			instance := infra.NewBlockCollector(100)
			signal := make(chan struct{})
			var wg sync.WaitGroup
			for i := 0; i < 100; i++ {
				wg.Add(1)
				go func() {
					defer wg.Done()

					if instance.Commit(uint64(1)) {
						close(signal)
					}
				}()
			}

			Expect(signal).To(BeClosed())
		})
	})
})
```
Glad you covered parallel commit case. Some suggestions:
- test api, not implementation details: test should not be modified during refactor
- avoid sleep in test

File: pkg/infra/observer.go [hunk 9]
```diff
@@ -51,9 +75,15 @@ func (o *Observer) Start(N int, errorCh chan error, finishCh chan struct{}, now
 			errorCh <- errors.Errorf("received nil message, but expect a valid block instead. You could look into your peer logs for more info")
 			return
 		}
-
+		// receive block
 		fb := r.Type.(*peer.DeliverResponse_FilteredBlock)
+		// logging
+		fmt.Printf("receivedTime %8.2fs\tBlock %6d\tTx %6d\t Address %s\n", time.Since(now).Seconds(), fb.FilteredBlock.Number, len(fb.FilteredBlock.FilteredTransactions), o.Address)
```
Comment (guoger) [inline pkg/infra/observer.go:81]: shouldn't this be printed via logger, instead of `fmt.Printf`?

File: pkg/infra/proposer.go [hunk 10]
```diff
@@ -33,11 +30,12 @@ func CreateProposers(conn, client int, nodes []Node, logger *log.Logger) (*Propo
...
 	}
 
-	return &Proposers{workers: ps, client: client, logger: logger}, nil
+	return &Proposers{workers: ps}, nil
 }
 
-func (ps *Proposers) Start(signed []chan *Elements, processed chan *Elements, done <-chan struct{}, config Config) {
-	ps.logger.Infof("Start sending transactions.")
+func (ps *Proposers) Start(signed []chan *Elements, processed chan *Elements, done <-chan struct{}, config Config, logger *log.Logger) {
+	// TODO global logger
```
Comment (guoger) [inline pkg/infra/proposer.go:37]: why exactly?

File: pkg/infra/benchmark_test.go [hunk 11]
```diff
@@ -60,3 +60,31 @@ func BenchmarkPeerEndorsement1(b *testing.B) { benchmarkNPeer(1, b) }
 func BenchmarkPeerEndorsement2(b *testing.B) { benchmarkNPeer(2, b) }
 func BenchmarkPeerEndorsement4(b *testing.B) { benchmarkNPeer(4, b) }
 func BenchmarkPeerEndorsement8(b *testing.B) { benchmarkNPeer(8, b) }
+
+func benchmarkCollector(concurrent int, b *testing.B) {
```
Comment (guoger) [inline pkg/infra/benchmark_test.go:64]: glad to see benchmarks :)

File: pkg/infra/benchmark_test.go [hunk 12]
```diff
@@ -60,3 +60,31 @@ func BenchmarkPeerEndorsement1(b *testing.B) { benchmarkNPeer(1, b) }
...
+	instance := NewBlockCollector(concurrent)
+	processed := make(chan bool, b.N)
```
Comment (guoger) [inline pkg/infra/benchmark_test.go:66]: not a big deal but this could be `chan struct{}`

**Review by guoger — CHANGES_REQUESTED** [review]

File: pkg/infra/BlockCollector.go (hunk 7 above)
Comment (SamYuan1990) [reply pkg/infra/BlockCollector.go:42]: thx

File: pkg/infra/block_collector.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
+package infra
+
+import "sync"
+
+type BlockCollector struct {
+	sync.Mutex
+	threshold int
+	registry  map[uint64]int
```
Comment (SamYuan1990) [inline pkg/infra/block_collector.go:8]: here, need to upgrade.
we have 3 peers and threshold equals 2.

File: pkg/infra/block_collector.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	// read map
+	remaining, ok := bc.registry[block]
+	if !ok {
+		bc.registry[block] = bc.threshold - 1
+		return false
+	}
+
+	if remaining == 1 {
+		delete(bc.registry, block)
+		return true
```
Comment (SamYuan1990) [inline pkg/infra/block_collector.go:34]: when peer1 and peer2 completed the process.
here comes peer3. which never reach here and memory leak?

File: pkg/infra/block_collector.go [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+	ci.Lock()
+	defer ci.Unlock()
+	ci.Addrs = append(ci.Addrs, addr)
+}
+
+func (ci *CollectionInfo) Len() int {
+	ci.Lock()
+	defer ci.Unlock()
+	return len(ci.Addrs)
+}
```
Comment (guoger) [inline pkg/infra/block_collector.go:26]: why do we need this?

File: pkg/infra/block_collector.go [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+type BlockCollector struct {
+	sync.Mutex
+	threshold  int
+	totalAddrs int
+	registry   map[uint64]*CollectionInfo
+}
+
+func NewBlockCollector(threshold int, totalAddrs int) *BlockCollector {
```
Comment (guoger) [inline pkg/infra/block_collector.go:35]: `(threshold, total int)`

File: pkg/infra/block_collector.go [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+func (ci *CollectionInfo) Len() int {
+	ci.Lock()
+	defer ci.Unlock()
+	return len(ci.Addrs)
+}
+
+type BlockCollector struct {
+	sync.Mutex
+	threshold  int
+	totalAddrs int
```
Comment (guoger) [inline pkg/infra/block_collector.go:31]: totalAddrs -> total

File: pkg/infra/block_collector_test.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+var _ = Describe("BlockCollector", func() {
+
+	Context("Commit", func() {
+		It("should return true when observer is 1", func() {
+			instance := infra.NewBlockCollector(1, 1)
+			Expect(instance.Commit(uint64(1), "addr")).To(BeTrue())
+		})
+
+		It("Should added when the block committed with threshold at 1st time", func() {
```
Comment (guoger) [inline pkg/infra/block_collector_test.go:19]: this sentence doesn't read right... and it's exactly the same as the test case below

File: pkg/infra/block_collector_test.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+			instance := infra.NewBlockCollector(2, 3)
+			Expect(instance.Commit(uint64(1), "addr1")).To(BeFalse())
+			Expect(instance.Commit(uint64(1), "addr2")).To(BeTrue())
+			Expect(instance.Commit(uint64(1), "addr3")).To(BeFalse())
+		})
+
+		It("Should supports multiple running", func() {
```
Comment (guoger) [inline pkg/infra/block_collector_test.go:33]: should support parallel committers

**Review by guoger — CHANGES_REQUESTED** [review]

File: pkg/infra/block_collector.go [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+		}
+		// all peers returned avoid save
+		if committedCount+1 == bc.total {
+			delete(bc.registry, block)
+		} else {
+			bc.registry[block] = committedCount + 1
+		}
+	}
+	return rs
+}
```
Comment (guoger) [inline pkg/infra/block_collector.go:46]:
```go
func NewBlockCollector(threshold int, total int) *BlockCollector {
	return &BlockCollector{
		threshold: threshold,
		total:     total,
		registry:  make(map[uint64]int),
	}
}

func (bc *BlockCollector) Commit(block uint64) (committed bool) {
	if bc.threshold == 1 {
		return true
	}
	bc.Lock()
	defer bc.Unlock()

	// read map
	committedCount, ok := bc.registry[block]
	if !ok {
		bc.registry[block] = 1
		return false
	}

	committedCount++
	// match the threshold
	if committedCount == bc.threshold {
		committed = true
	}

	// all peers returned avoid save
	if committedCount == bc.total {
		delete(bc.registry, block)
	} else {
		bc.registry[block] = committedCount
	}

	return
}
```

File: pkg/infra/block_collector.go [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+	}
+}
+
+func (bc *BlockCollector) Commit(block uint64, addr string) (committed bool) {
+	if bc.threshold == 1 {
+		return true
+	}
+	bc.Lock()
+	defer bc.Unlock()
+	rs := false
```
Comment (guoger) [inline pkg/infra/block_collector.go:27]: you could use named return value `committed` here

File: pkg/infra/block_collector.go [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+func NewBlockCollector(threshold int, total int) *BlockCollector {
+	registry := make(map[uint64]int)
+	return &BlockCollector{
+		threshold: threshold,
+		total:     total,
+		registry:  registry,
+	}
+}
+
+func (bc *BlockCollector) Commit(block uint64, addr string) (committed bool) {
```
Comment (guoger) [inline pkg/infra/block_collector.go:21]: `addr string` is not used anywhere, pls remove

File: e2e/util.go [hunk 23]
```diff
@@ -75,7 +79,89 @@ sign_cert: {{.SignCert}}
...
+`
+	var err error
+	var tmpl *template.Template
+	tmpl, err = template.New("test").Parse(Text)
+	if values.Peers == 2 {
+		tmpl, err = template.New("test").Parse(Text2)
+	}
+	if values.Peers == 3 {
+		tmpl, err = template.New("test").Parse(Text3)
+	}
```
Comment (guoger) [inline e2e/util.go:164]: Go template is capable to handle array, so you don't need to write separate templates specifically for 2&3

File: e2e/e2e_test.go [hunk 24]
```diff
@@ -332,5 +348,90 @@ var _ = Describe("Mock test", func() {
 				Eventually(tapeSession.Out).Should(Say("Time.*Block.*Tx.*10.*"))
 			})
 		})
+
+		When("Multiple committers", func() {
```
Comment (guoger) [inline e2e/e2e_test.go:352]: Please write more descriptive text for test cases - they should describe what actually is being done during test

File: pkg/infra/block_collector.go (hunk 20 above)
Comment (guoger) [reply pkg/infra/block_collector.go:46]: Also, maybe we should add a check for `threshold > total` within constructor?

File: pkg/infra/block_collector_test.go [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"sync"
+	"tape/pkg/infra"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("BlockCollector", func() {
```
Comment (guoger) [inline pkg/infra/block_collector_test.go:11]: i think we should have at least following cases:
```
threshold = 1, total = 1
threshold = 1, total = 2
threshold = 2, total = 4
threshold = 4, total = 4
threshold = 2, total = 1 (error)
```

**Review by guoger — CHANGES_REQUESTED** [review]

**Review by guoger — APPROVED** [review]
